lib/mastodon/cli/domains.rb

```diff
@@ -4,7 +4,90 @@
 require_relative 'base'
 
 module Mastodon::CLI
+  # This class is a helper so we can reuse the validation logic in
+  # Admin::Import, which typically expects ActionDispatch::Parameters as an
+  # argument
+  class DomainBlocklistData
+    attr_accessor :original_filename, :path
+
+    def self.from(filename, filepath)
+      import_data = new
+      import_data.original_filename = filename
+      import_data.path = filepath
+      import_data
+    end
+  end
+
+  class DomainBlocklist < Base
+    desc 'import [NAME] [URL]', 'Imports a blocklist from the given URL'
+    def import(name, url)
+      filename = "#{name}.csv"
+      tmpfile = Tempfile.new(filename)
+
+      begin
+        say('Fetching blocklist...')
+        response = HTTP.headers(
+          'Accept' => 'text/csv',
+          'User-Agent' => Mastodon::Version.user_agent
+        ).get(url)
+
+        unless response.code == 200
+          say("Error: Failed to fetch blocklist from: #{url}", :red)
+          exit(1)
+        end
+
+        tmpfile.write(response.body.to_s)
+
+        say('Fetched blocklist!')
+
+        import = Admin::Import.new(data: DomainBlocklistData.from(filename, tmpfile.path))
+
+        unless import.validate
+          import_errors = import.errors.full_messages.map { |err| "- #{err}" }.join("\n")
+
+          say("Error: Failed to import blocklist, details:\n#{import_errors}", :red)
+          exit(1)
+        end
+
+        say("Importing #{import.csv_rows.count} domain blocks... (this may take a while)")
+
+        global_private_comment = "Initial import of #{name} blocklist"
+        domain_blocks = import.csv_rows.filter_map do |row|
+          domain = row['#domain'].strip
+          next if DomainBlock.rule_for(domain).present?
+
+          domain_block = DomainBlock.new(domain: domain,
+                                         severity: row.fetch('#severity', :suspend),
+                                         reject_media: row.fetch('#reject_media', false),
+                                         reject_reports: row.fetch('#reject_reports', false),
+                                         private_comment: global_private_comment,
+                                         public_comment: row['#public_comment'],
+                                         obfuscate: row.fetch('#obfuscate', false))
+
+          if domain_block.invalid?
+            say("Error importing block: #{domain}, #{domain_block.errors.full_messages.join(', ')}", :red)
+            next
+          end
+
+          domain_block.save!
+          DomainBlockWorker.perform_async(domain_block.id)
```

Comment:
Note: we're not adding to the AdminLog here, as we may not have a user account yet, so there'd be no account to attribute this import to.

Reply:
Arguably we only need DomainBlockWorker if the server has already been running / has peers, I think!